libp2p/NodeTable.cpp

```diff
@@ -0,0 +1,466 @@
+		});
+
+		if (s.nodes.size() >= s_bucketSize)
+		{
+			contested = s.nodes.front().lock();
+			if (!contested)
+			{
+				s.nodes.pop_front();
+				s.nodes.push_back(_n);
+			}
+		}
+		else
+			s.nodes.push_back(_n);
+	}
+	
+	if (contested)
+		evict(contested, _n);
+}
+
+void NodeTable::dropNode(std::shared_ptr<NodeEntry> _n)
+{
+	NodeBucket &s = bucket(_n.get());
+	{
+		Guard l(x_state);
+		s.nodes.remove_if([&_n](std::weak_ptr<NodeEntry> n) { return n.lock() == _n; });
+	}
+	Guard l(x_nodes);
+	m_nodes.erase(_n->id);
+}
+
+NodeTable::NodeBucket& NodeTable::bucket(NodeEntry const* _n)
+{
+	return m_state[_n->distance - 1];
+}
+
+void NodeTable::onReceived(UDPSocketFace*, bi::udp::endpoint const& _from, bytesConstRef _packet)
+{
+	// h256 + Signature + RLP (smallest possible packet is empty neighbors packet which is 3 bytes)
+	if (_packet.size() < h256::size + Signature::size + 3)
+	{
+		clog(NodeTableMessageSummary) << "Invalid Message size from " << _from.address().to_string() << ":" << _from.port();
+		return;
+	}
+	
+	bytesConstRef signedBytes(_packet.cropped(h256::size, _packet.size() - h256::size));
+	h256 hashSigned(sha3(signedBytes));
+	if (!_packet.cropped(0, h256::size).contentsEqual(hashSigned.asBytes()))
+	{
+		clog(NodeTableMessageSummary) << "Invalid Message hash from " << _from.address().to_string() << ":" << _from.port();
+		return;
+	}
+
+	bytesConstRef rlpBytes(signedBytes.cropped(Signature::size, signedBytes.size() - Signature::size));
+	RLP rlp(rlpBytes);
+	unsigned itemCount = rlp.itemCount();
+	
+	bytesConstRef sigBytes(_packet.cropped(h256::size, Signature::size));
+	Public nodeid(dev::recover(*(Signature const*)sigBytes.data(), sha3(rlpBytes)));
+	if (!nodeid)
+	{
+		clog(NodeTableMessageSummary) << "Invalid Message signature from " << _from.address().to_string() << ":" << _from.port();
+		return;
+	}
+	noteNode(nodeid, _from);
+	
+	try {
+		switch (itemCount) {
+			case 1:
+			{
+//				clog(NodeTableMessageSummary) << "Received Pong from " << _from.address().to_string() << ":" << _from.port();
+				Pong in = Pong::fromBytesConstRef(_from, rlpBytes);
+				
+				// whenever a pong is received, first check if it's in m_evictions
+				
+				break;
+			}
+				
+			case 2:
+				if (rlp[0].isList())
+				{
+					Neighbors in = Neighbors::fromBytesConstRef(_from, rlpBytes);
+//					clog(NodeTableMessageSummary) << "Received " << in.nodes.size() << " Neighbors from " << _from.address().to_string() << ":" << _from.port();
+					for (auto n: in.nodes)
+						noteNode(n.node, bi::udp::endpoint(bi::address::from_string(n.ipAddress), n.port));
+				}
+				else
+				{
+//					clog(NodeTableMessageSummary) << "Received FindNode from " << _from.address().to_string() << ":" << _from.port();
+					FindNode in = FindNode::fromBytesConstRef(_from, rlpBytes);
+					
+					std::vector<std::shared_ptr<NodeTable::NodeEntry>> nearest = findNearest(in.target);
+					static unsigned const nlimit = (m_socketPtr->maxDatagramSize - 11) / 86;
```

Comment:
Maybe define 11 and 86 as either macros or `const int` somewhere? It will help the reader understand what they are.